src/actions/requests.rs

```diff
@@ -618,37 +636,54 @@ impl<'a> Action<'a> for CodeAction {
 
 impl<'a> RequestAction<'a> for CodeAction {
     type Response = Vec<Command>;
-    fn handle<O: Output>(&mut self, _id: usize, params: Self::Params, ctx: &mut ActionContext, _out: O) -> Result<Self::Response, ()> {
+    fn handle<O: Output>(&mut self, _id: usize, params: Self::Params, ctx_: &mut ActionContext, _out: O) -> Result<Self::Response, ()> {
         trace!("code_action {:?}", params);
 
-        let ctx = ctx.inited();
+        let ctx = ctx_.inited();
         let file_path = parse_file_path!(&params.text_document.uri, "code_action")?;
 
-        match ctx.previous_build_results.lock().unwrap().get(&file_path) {
-            Some(ref diagnostics) => {
-                let suggestions = diagnostics.iter().filter(|&&(ref d, _)| d.range == params.range).flat_map(|&(_, ref ss)| ss.iter());
-                let mut cmds = vec![];
-                for s in suggestions {
-                    let span = Location {
-                        uri: params.text_document.uri.clone(),
-                        range: s.range,
-                    };
-                    let span = serde_json::to_value(&span).unwrap();
-                    let new_text = serde_json::to_value(&s.new_text).unwrap();
+        let mut cmds = vec![];
+
+        // search for compiler suggestions
+        if let Some(diagnostics) = ctx.previous_build_results.lock().unwrap().get(&file_path) {
+            let suggestions = diagnostics.iter().filter(|&&(ref d, _)| d.range == params.range).flat_map(|&(_, ref ss)| ss.iter());
```

Comment:
Could you factor this code (and the deglob code) into its own function, so that the handle function is only responsible for deciding what suggestions respond with and compiling them into a result, but not actually making the suggestions?

Reply:
Sounds good